lib/debugger.ts

```diff
@@ -0,0 +1,273 @@
+// Util from NodeJs
+import * as net from 'net';
+import * as util from 'util'
+
+import {ProtractorBrowser} from './browser';
+import {Locator} from './locators';
+import {Logger} from './logger';
+import {Ptor} from './ptor';
+import * as helper from './util';
+
+declare var global: any;
+declare var process: any;
+
+let logger = new Logger('protractor');
+let webdriver = require('selenium-webdriver');
+
+export class DebugHelper {
+  browser: ProtractorBrowser;
+
+  /**
+   * Set to true when we validate that the debug port is open. Since the debug
+   * port is held open forever once the debugger is attached, it's important
+   * we only do validation once.
+   */
+  debuggerValidated_: boolean;
+
+  dbgCodeExecutor: any;
+
+  constructor(browser: ProtractorBrowser) {
+    this.browser = browser;
+  }
+
+  init(
+      debuggerClientPath: string, onStartFn: Function, opt_debugPort?: number) {
+    webdriver.promise.ControlFlow.prototype.getControlFlowText = function() {
+      let controlFlowText = this.getSchedule(/* opt_includeStackTraces */ true);
+      // This filters the entire control flow text, not just the stack trace, so
+      // unless we maintain a good (i.e. non-generic) set of keywords in
+      // STACK_SUBSTRINGS_TO_FILTER, we run the risk of filtering out non stack
+      // trace. The alternative though, which is to reimplement
+      // webdriver.promise.ControlFlow.prototype.getSchedule() here is much
+      // hackier, and involves messing with the control flow's internals /
+      // private variables.
+      return helper.filterStackTrace(controlFlowText);
+    };
+
+    let vm_ = require('vm');
+    let flow = webdriver.promise.controlFlow();
+
+    interface Context {
+      require: any;
+      [key: string]: any;
+    }
+    let context: Context = {require: require};
+    global.list = (locator: Locator) => {
+      return (<Ptor>global.protractor)
```

Comment:
Fixed where possible. We have to leave things the way they are for dbgCodeExecutor, but I can fix that when I untangle it later.